diff --git a/drivers/mmc/core/bus.c b/drivers/mmc/core/bus.c
index fb2ca3d75084..52fe79de5001 100644
--- a/drivers/mmc/core/bus.c
+++ b/drivers/mmc/core/bus.c
@@ -167,19 +167,9 @@ static int mmc_bus_suspend(struct device *dev)
 	if (mmc_bus_needs_resume(host))
 		return 0;
 	ret = host->bus_ops->suspend(host);
-
-	/*
-	 * bus_ops->suspend may fail due to some reason
-	 * In such cases if we return error to PM framework
-	 * from here without calling pm_generic_resume then mmc
-	 * request may get stuck since PM framework will assume
-	 * that mmc bus is not suspended (because of error) and
-	 * it won't call resume again.
-	 *
-	 * So in case of error call pm_generic_resume().
-	 */
 	if (ret)
 		pm_generic_resume(dev);
+
 	return ret;
 }
 
