diff --git a/cmake/Docs.cmake b/cmake/Docs.cmake
index 2a3faf896..bacd3296a 100644
--- a/cmake/Docs.cmake
+++ b/cmake/Docs.cmake
@@ -18,7 +18,7 @@ add_custom_target(sphinx-docs
     mkdir -p ${SPHINX_HTML_DIR}/_static/
     COMMAND env PATH="${CMAKE_BINARY_DIR}:$$PATH"
         ${SPHINX_EXECUTABLE}
-        -j auto
+        -j 1
         -q -b html
         -c "${SPHINX_SRC_DIR}"
         -d "${SPHINX_ROOT_DIR}/.doctrees-html"
