diff --git a/include/linux/compiler-clang.h b/include/linux/compiler-clang.h
index 7957637835e3..de179993e039 100644
--- a/include/linux/compiler-clang.h
+++ b/include/linux/compiler-clang.h
@@ -11,10 +11,7 @@
 #define uninitialized_var(x) x = *(&(x))
 #endif
 
-/*
-* GCC does not warn about unused static inline functions for
-* -Wunused-function.  This turns out to avoid the need for complex #ifdef
-* directives.  Suppress the warning in clang as well.
-*/
-#undef inline
-#define inline inline __attribute__((unused)) notrace
+/* same as gcc, this was present in clang-2.6 so we can assume it works
+ * with any version that can compile the kernel
+ */
+#define __UNIQUE_ID(prefix) __PASTE(__PASTE(__UNIQUE_ID_, prefix), __COUNTER__)
