diff --git a/.git-prompt.sh b/.git-prompt.sh
index 983e419d2b..7216682220 100644
--- a/.git-prompt.sh
+++ b/.git-prompt.sh
@@ -195,13 +195,13 @@ __git_ps1_show_upstream ()
 		"") # no upstream
 			p="" ;;
 		"0	0") # equal to upstream
-			p="=" ;;
+			p="" ;;
 		"0	"*) # ahead of upstream
-			p=">" ;;
+			p=" >" ;;
 		*"	0") # behind upstream
-			p="<" ;;
+			p=" <" ;;
 		*)	    # diverged from upstream
-			p="<>" ;;
+			p=" <>" ;;
 		esac
 	else
 		case "$count" in
@@ -453,8 +453,7 @@ __git_ps1 ()
 					git describe --tags --exact-match HEAD ;;
 				esac 2>/dev/null)" ||
 
-				b="$short_sha..."
-				b="($b)"
+				b="$short_sha"
 			fi
 		fi
 	fi
