diff --git a/fs/nilfs2/inode.c b/fs/nilfs2/inode.c
index c1219c0678a5..476a4a649f38 100644
--- a/fs/nilfs2/inode.c
+++ b/fs/nilfs2/inode.c
@@ -204,7 +204,7 @@ static bool nilfs_dirty_folio(struct address_space *mapping,
 {
 	struct inode *inode = mapping->host;
 	struct buffer_head *head;
-	unsigned int nr_dirty;
+	unsigned int nr_dirty = 0;
 	bool ret = filemap_dirty_folio(mapping, folio);
 
 	/*
@@ -215,7 +215,6 @@ static bool nilfs_dirty_folio(struct address_space *mapping,
 	if (head) {
 		struct buffer_head *bh = head;
 
-		nr_dirty = 0;
 		do {
 			/* Do not mark hole blocks dirty */
 			if (buffer_dirty(bh) || !buffer_mapped(bh))
@@ -225,7 +224,7 @@ static bool nilfs_dirty_folio(struct address_space *mapping,
 			nr_dirty++;
 		} while (bh = bh->b_this_page, bh != head);
 	} else if (ret) {
-		nr_dirty = 1 << (PAGE_SHIFT - inode->i_blkbits);
+		nr_dirty = 1 << (folio_shift(folio) - inode->i_blkbits);
 	}
 	spin_unlock(&mapping->private_lock);
 
