diff --git a/include/linux/compiler-gcc.h b/include/linux/compiler-gcc.h
index 143d40e8a1ea..3104e5d352b7 100644
--- a/include/linux/compiler-gcc.h
+++ b/include/linux/compiler-gcc.h
@@ -302,3 +302,16 @@
  * code
  */
 #define uninitialized_var(x) x = x
+
+#if GCC_VERSION >= 50100
+#define COMPILER_HAS_GENERIC_BUILTIN_OVERFLOW 1
+#endif
+
+/*
+ * Show that FORTIFY_SOURCE is not supported by GCC 4.9.4 and earlier
+ */
+#if GCC_VERSION <= 40904
+#ifndef __NO_FORTIFY
+#define __NO_FORTIFY 1
+#endif
+#endif
